PMON High Level Design for SmartSwitch  (Replaces the PR #1538)

rameshraghupathy opened the pull request:
PMON High Level Design for SmartSwitch  (Replaces  the initial PR #1538  https://github.com/sonic-net/SONiC/pull/1538)

The following 4 PRs have been raised for PMON Platform Independent changes:
https://github.com/sonic-net/sonic-buildimage/pull/18676
https://github.com/sonic-net/sonic-platform-daemons/pull/467
https://github.com/sonic-net/sonic-platform-common/pull/454
https://github.com/sonic-net/sonic-utilities/pull/3271

prgeor commented on doc/smart-switch/pmon/static-ip-assignment.md:
```diff
@@ -0,0 +1,51 @@
+# Midplane Interface 
```
@rameshraghupathy can we remove this since smartswitch advocates for DHCP method for IP assignment?

rameshraghupathy replied:
Done

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,495 @@
+    * Solution 2. static ip allocation based on DPU ID. [link](https://github.com/rameshraghupathy/SONiC/blob/origin/ss_pmon_hld/doc/smart-switch/pmon/static-ip-assignment.md)
+* For midplane-interface IP address allocation we will follow the procedure in solution.1 [link](https://github.com/sonic-net/SONiC/blob/master/doc/smart-switch/ip-address-assigment/smart-switch-ip-address-assignment.md)
+
+### 3.3.1.    MAC address distribution
```
@rameshraghupathy can we remove this as this is platform specific detail

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,495 @@
+* When DPU card or the Supervisor boots and as part of its initialization, midplane interface gets initialized.
+* Two solutions were proposed for this purpose.  
+    * Solution 1. uses a dhcp server - client model between the switch host and the dpus to allocate IP address. [link](https://github.com/sonic-net/SONiC/blob/master/doc/smart-switch/ip-address-assigment/smart-switch-ip-address-assignment.md)
+    * Solution 2. static ip allocation based on DPU ID. [link](https://github.com/rameshraghupathy/SONiC/blob/origin/ss_pmon_hld/doc/smart-switch/pmon/static-ip-assignment.md)
```
@rameshraghupathy can we remove Solution 2.

rameshraghupathy replied:
Done

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,495 @@
+
+```
+
+show platform temperature - shows the DPU temperature on the switch
````
@rameshraghupathy Please indicate (preferably at the top right corner in red ink) where this CLI is executed: DPU or NPU

rameshraghupathy replied:
Done

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,495 @@
+
+show platform temperature - shows the DPU temperature on the switch
+```
+root@sonic:~#show platform temperature
````
@rameshraghupathy Please indicate preferably at the top right corner in red ink where this CLI is executed: DPU or NPU

rameshraghupathy replied:
Highlighted the location of execution of the CLI.

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,495 @@
+
+show platform fan - shows the fan speed and status on the switch
+```
+root@sonic:~#show platform fan
````
@rameshraghupathy Please indicate preferably at the top right corner in red ink where this CLI is executed: DPU or NPU

rameshraghupathy replied:
Done

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,495 @@
+```
+show reboot-cause history on the switch shows the dpu reboot-cause as well.
+```
+root@sonic:~#show reboot-cause history
````
@rameshraghupathy Please indicate preferably at the top right corner in red ink where this CLI is executed: DPU or NPU or both

rameshraghupathy replied:
Highlighted the location of execution of the CLI.

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,495 @@
+
+switch          2023_10_20_18_52_28     Watchdog:1 expired;         Wed 20 Oct 2023 06:52:28 PM UTC     N/A     N/A
+switch          2023_10_05_18_23_46     reboot                      Wed 05 Oct 2023 06:23:46 PM UTC     user    N/A
+DPU0            2023_10_04_18_23_46     Power Loss                  Tue 04 Oct 2023 06:23:46 PM UTC     N/A     N/A
+DPU3            2023_10_03_18_23_46     Watchdog: stage 1 expired;  Mon 03 Oct 2023 06:23:46 PM UTC     N/A     N/A
+DPU3            2023_10_02_18_23_46     Host Power-cycle            Sun 02 Oct 2023 06:23:46 PM UTC     N/A     Host lost DPU
+DPU3            2023_10_02_17_23_46     Host Reset DPU              Sun 02 Oct 2023 05:23:46 PM UTC     N/A     N/A
+DPU2            2023_10_02_17_20_46     reboot                      Sun 02 Oct 2023 05:20:46 PM UTC     admin   User issued 'reboot'
```
@rameshraghupathy can you elaborate how NPU is able to fetch DPU reboot cause. Can we have the DB schema for this ?

rameshraghupathy replied:
Done

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,495 @@
+
+}
+```
+#### 3.1.4 ModuleBase class API enhancements
````
@rameshraghupathy can you remove `HW mgmt` column as this is platform specific information

rameshraghupathy replied:
Done

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,495 @@
+}
+```
+#### 3.1.4 ModuleBase class API enhancements
+| API | current_usage | switch cpu | dpu | SONiC enhancements | PD plugin change| HW mgmt change | comments |
````
@rameshraghupathy I would suggest instead of this table please define the API name, paramaters and the return. We only need to consider APIs which are going to be new and if any change to existing API

rameshraghupathy replied:
Done

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,495 @@
+| get_module_dpu_port(self, index) | Retrieves the DPU port (internal ASIC port for DPU) represented by DPU index - 1,8 | yes | na | yes | see comments | Platforms that require to overwrite the platform.json file will use this API. See section 3.1.3 below |
+
+#### 3.1.3 NPU to DPU port mapping
+platform.json of NPU/switch will show the NPU port to DPU port mapping. This will be used by services early in the system boot for midplane IP assignment. In this example there are 8 DPUs and ach having a 200G interface.
+```
+"DPUs" : [
+    {
+      "DPU0": {
+                "Ethernet224": "Ethernet0/0"
+       }
+    },
+    {
+       "DPU1": {
+                "Ethernet228": "Ethernet1/0"
+        },
+    },
+    .
+    .
+    {
+       "DPU7": {
+                "Ethernet252": "Ethernet7/0"
+        },
+    },
````
@rameshraghupathy the data ports are not used for midplane IP assignments. Please check.

rameshraghupathy replied:
This was a mistake.  Fixed it.

oleksandrivantsiv commented on doc/smart-switch/pmon/smartswitch-pmon.md.save:
```diff
@@ -0,0 +1,495 @@
+# SmartSwitch PMON High Level Design
```
Do we need this file?

rameshraghupathy replied:
Thanks for pointing! Cleaned.

oleksandrivantsiv commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,806 @@
+
+## 2.	Requirements and Assumptions
+### 2.1.    Onboarding
+* The SmartSwitch host PMON should be able to Power Cycle, Shutdown, Reset, and rest the PCIe link per DPU or the entire system
```
We should have a config DB table that will represent an admin state of the DPU. For example, if we want to use a smart switch as a switch only we should have the ability to power off the DPUs and have this as a permanent configuration. This can reduce the power consumption of the box.

I see that there is a `set_admin_state(...)` API defined. But without having the configuration in the config DB any changes to the DPU admin state will have a temporary effect.

rameshraghupathy replied:
As per the requirement the smartswitch will boot up as a switch by default, meaning that all the DPUs will remain in power down state until unless explicitly booted.  Explained this in detail. Don't see a need for config as this is the default behavior.  Let us keep it simple.

oleksandrivantsiv replied:
With the proposed design there is no possibility to make DPU admin state permanent. If the DPU by default is in the down state during the system start who is responsible for bringing it up? If the user sets the DPU to the up state and then does the power cycle of the smart switch what will be the state of the DPU after the smart switch boot?

The assumption that a smart switch boots as a switch by default is based on the behavior of one specific platform. The behavior might be different on the other platforms. That is why we need to have a DPU state in the config DB.

oleksandrivantsiv commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,806 @@
+      True
+```
+
+get_module_dpu_port(self, index):
````
There are at least two ports connected to the DPU. One midplane (PCI) interface and one or more data plane interfaces. If I understand correctly, this API should return the mapping from the data plane port on the NPU to the data plane port on the DPU, e.g.: {"Ethernet224": "Ethernet0"}

oleksandrivantsiv commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,806 @@
+        A bool value, should return True if module is reachable via midplane
+```
+
+get_all_asics(self):
````
What is the use case of this API?

rameshraghupathy replied:
This is an existing API for modular chassis. A DPU card may or may not have other ASICs.  This may be applicable to those cases.  We don't see a need. If you agree we can remove this.  So, I'll remove this.

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,281 @@
+SCHEMA
+key:  dpu_state:1
+ 
+HMSET dpu_state:1
```
last time when we talked, we had decided to to keep these details as vendor specific, and this can be part of health details

rameshraghupathy replied:
Outdated

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,281 @@
+    * Once DPU boots it will update the remaining DPU_STATE information in the SmartSwitch ChassisStateDB
+* APIs
+    * Existing APIs: get_name, get_type, get_my_slot_idx, etc
+    * New APIs
```
use these existing APIs
- get_oper_status(self)
- get_module_index(self, module_name)
- get_midplane_ip(self):
- get_base_mac(self)

Note: there is no notion of slot for DPU, rather use port number

rameshraghupathy replied:
Outdated

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,281 @@
+
+<p align="center"><img src="./images/pmon-vertical.svg"></p>
+
+## 2.	Requirements, Assumptions and SLA
```
I would suggest what is new requirements vs. current chassis model. Goal is we should be able to reuse most of functionalities. For example - DPU life cycle management should be similar to Line Card in chassis based system

rameshraghupathy replied:
Outdated

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,281 @@
+* Reboot
+    * Only cold reboot of DPUs is required, warm boot support is not required.
+### 2.2. Monitoring and Thermal Management
+* Dpu State
```
we should keep this similar to Line Card state. However, additional sub states can be vendor specific debug state, which can be exposed in health state

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,281 @@
+        ```
+### 2.3. Detect and Debug
+* Health
+    * SmartSwitch DPUs should store their health data in their local StateDB 
````
we should have a summary health state on chassis db

rameshraghupathy replied:
A decision was made not to store this data on the chassisStateDB as this is high volume data.

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,281 @@
+    * SmartSwitch DPUs should store their health data in their local StateDB 
+    * DPUs should support a CLI to display the health data “show dpu health”
+    * The host should be able to access this data using a redis call or an api
+* Alarm and Syslog
```
what is new?

rameshraghupathy replied:
Nothing new in general. It is a natural extension for smartwitch HW.

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,281 @@
+* changes to platform plugin
+    * extend module.py to support dpu module
+        * class DpuCardModule(Module):
+    * Update the DPU_STATE table information read from platform (refer DB schema section)
```
we should keep it similar to line card state

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,281 @@
+* Has reason for previous reboot
+* Carries DP, CP up down reasons
+
+DPU_STATE Table
```
provide this as part of vendor specific health state

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,281 @@
+| 7 | 169.254.7.1 | 169.254.7.2 | DPU_MP_MAC7 | HOST_MP_MAC7 |
+| 8 | 169.254.8.1 | 169.254.8.2 | DPU_MP_MAC8 | HOST_MP_MAC8 |
+
+* The IP address allocation for the DPUs and the host can be static as in Modular chassis design 
```
no DHCP?

rameshraghupathy replied:
Looks like you are looking at an older version of the PR.  Please refer to the recent version

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,806 @@
+        A string, providing the vendor's product description of the module.
+```
+
+get_slot(self):
````
not applicable for smart switch

rameshraghupathy replied:
Fixed the description: Retrieves the DPU ID in case of smartswitch

prgeor replied:
@rameshraghupathy we discussed to remove slot concept from Smart swtich. Instead have new API get_dpu_id()

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,806 @@
+    SCHEMA
+    key:  dpu_state:1
+    
+    HMSET dpu_state:1
```
These details being on DB is overkill. These can be in some debug log

rameshraghupathy replied:
This is the KEY for the DB. Doesn't signify the DPU STATE.

manapalai commented on doc/smart-switch/pmon/static-ip-assignment.md:
```diff
@@ -0,0 +1,51 @@
+* When DPU card (SLED) or the Supervisor boots and as part of its initialization, midplane interface gets initialized.
+* Slot number is used in assigning an IP address to these interfaces.
+
+## 1.    DPU static IP address allocation
```
our POR is DHCP, right?

rameshraghupathy replied:
Yes is the answer, but this is outdated.

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,806 @@
+Ethernet216  1032,1033,1034,1035,1036,1037,1038,1039     400G   9100    N/A    etp27  routed    down       up     N/A         N/A
+
+### SmartSwitch DPU0-X ###
+Ethernet224                          780,781,782,783     100G   9100    N/A   etp28a  routed    down       up     N/A         N/A
```
should 'Type' say these are internal DPU ports?

rameshraghupathy replied:
Sure, done

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,806 @@
+ ```
+ * Detailed output from the switch can be obtained with the following CLI
+ * The system health monitor list command should include the status of the DPU. If one or more DPUs are not ok it should be highlighted in the command output
+ * The switch will fetch the DPU information stored in the DPU's STATE_DB as needed
````
let us specify how (what transport) is used get DPU health

rameshraghupathy replied:
A gRPC client running on the switch PMON will fetch the data.  I'll updating the document.

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,806 @@
+root@sonic:~#show chassis modules status                                                                                      
+Name        Description         Physical-Slot       Oper-Status     Admin-Status    Serial
+
+DPU0        SS-DPU-0            1                   Online          up              SN20240105
```
let us make this description consistent e.g. DPU0 vs. DPU-0

rameshraghupathy replied:
Agree, Done

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,806 @@
+show chassis modules status - shows the dpu status of all DPUs and the Switch supervisor     <font>**`Executed on the switch`**</font></p>
+```
+root@sonic:~#show chassis modules status                                                                                      
+Name        Description         Physical-Slot       Oper-Status     Admin-Status    Serial
````
Physical-slot? Assume the data is here port number.

rameshraghupathy replied:
No it is not the physical slot number.  It represents the DPU ID and it is across this document.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,806 @@
+
+## 2.	Requirements and Assumptions
+### 2.1.    Onboarding
+* The SmartSwitch host PMON should be able to Power Cycle, Shutdown, Reset, and rest the PCIe link per DPU or the entire system
```
Done

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,806 @@
+    * The DPU must provide the state information once it boots its OS to DPU_STATE table.
+    * The SmartSwitch host PMON should be able to monitor the liveliness of the DPUs and when they go down should be able to take appropriate actions  and should try to gracefully recover the DPU when requested by the PMON
+
+* Thermal management
```
Agree, called out in the document.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,806 @@
+
+get_my_slot(self):
+```
+    Retrieves the slot number
````
Fixed the description: Retrieves the DPU ID in case of smartswitch

oleksandrivantsiv reviewed (changes requested):
I have several general comments on the design:
1. There is no definition of all CLI commands that should be used to control the DPUs.
2. There are no flow diagrams that describe relations between the CLI commands, databases, and the platform API. It is impossible to understand how PMON will work based on the available information in this document.
3. There is no sonic-mgmt test plan for the proposed new functionality. The test plan is required to understand the scope of the testing and coverage of the new functionality.

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,790 @@
+Ethernet248                                8,9,10,11     100G   9100    N/A   etp31a  routed    down       up     N/A         N/A
+Ethernet252                              12,13,14,15     100G   9100    N/A   etp31b  routed    down       up     N/A         N/A
+
+```
````
We should cover following functionalities, please add flows
- DPU power down/up flow - user can shut/unshut specific DPU
- DPU reboot
- Switch reboot

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,790 @@
+* The SmartSwitch host PMON should be able to Power Cycle, Shutdown, Reset, and rest the PCIe link per DPU or the entire system. The DPU_MODULE will behave like the LINE_CARD_MODULE with respect to these functions. 
+    #### SmartSwitch Expected PowerUp sequence:
+    * When the smartswitch device is booted, the host will boot first and leave the DPUs in power-down state by default.
+    * The user will powerup/boot the DPUs as needed.
```
really? we need user to power up?

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,790 @@
+
+    #### SmartSwitch Expected PowerDown sequence:
+    * When the smartswitch device is completely shutdown, it will try to gracefully shutdown all the DPUs first. In the event of failure to graceful shutdown will force shutdown.
+    * Then the switch will power down itself gracefully (NPU/Chassis)
```
this system flow is missing in later part of the doc

rameshraghupathy replied:
Power on/off sequence is addressed in the recent version.

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,790 @@
+    * The DPUs should provide their state to the host as the boot progression happens by updating the dpu state data in the DPU_STATE table in the host ChassisStateDB (explained in DB schema)
+    * DPUs should be able to store the data using a redis call
+    * The DPU must provide the state information once it boots its OS to DPU_STATE table.
+    * The SmartSwitch host PMON should be able to monitor the liveliness of the DPUs and when they go down should be able to take appropriate actions  and should try to gracefully recover the DPU when requested by the PMON
```
liveness: is this something new? if yes, please explain the flow

rameshraghupathy replied:
This has been there from the beginning.  This is the DPU_SATE data which can be used by a controller or a HA or LB logic to make decisions.  For example, if a DPU updates it data plane state to be down, the HA logic can do the cut over and inform the controller, upon receiving the cutover message the controller can inform the host PMON to reboot the DPU so that it is available for future use.

manapalai commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,790 @@
+      False
+```
+
+get_num_modules(self):
````
let us call out, these are existing APIs

rameshraghupathy replied:
Done

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,898 @@
+### 2.3. Detect and Debug
+* Health
+    * SmartSwitch DPUs should store their health data in their local StateDB 
+    * DPUs should support a CLI to display the health data “show chassis health-events”
```
@rameshraghupathy can we reuse existing `show system-health summary`

rameshraghupathy replied:
God find.  Forgot to clean it up.  It is cleaned up now.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,281 @@
+    * Extend existing CLIs such as show platform fan/temperature to support the new HW
+    * Add new CLIs 
+        ```
+        show dpu <id> state
````
outdated

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,281 @@
+        * get_dpu_mac (will take dpu ID or slot/offset as an argument)
+    * Update sensor, peripheral device date to sysfs, state-DB/ChassisStateDB (platform-DB)
+
+### 3.1. Thermal management
```
yes, but outdated

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,898 @@
+        MODULE_STATUS_FAULT, MODULE_STATUS_PRESENT or MODULE_STATUS_ONLINE
+```
+
+reboot(self, reboot_type):
````
@rameshraghupathy i am not sure of the use case. This looks to be very platform specific.

rameshraghupathy replied:
This is for booting each DPU module and it is platform common.

oleksandrivantsiv replied:
Agree with @prgeor. The `reboot_type` is very platform-specific. I suggest to remove this parameter

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,898 @@
+        A string, the IP-address of the module reachable over the midplane
+```
+
+is_midplane_reachable(self):
````
@rameshraghupathy For clarity I would suggest to create two sections
1. For new APIs (completely new)
2. Modification to existing API or their description or new addition

rameshraghupathy replied:
Already there are two sections for both ChassisBase and ModuleBase, one for existing API modifications and the other one for new APIs.

keboliu commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,947 @@
+#### Platform CLIs
+* Platform CLIs remain the same but include DPU, and changes in sensors and FRUs
+
+show platform inventory 
```
Is it a new command planned to be added in this HLD design? I didn't see it in the current SONiC code or I missed?

rameshraghupathy replied:
The common API is "show platform ...".  The extension is platform specific.  This is an example to enhance the corresponding platform API to display the DPU modules. Updated the doc to show that it is an example.

keboliu commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,947 @@
+```
+
+show platform fan - shows the fan speed and status on the switch       <font>**`Executed on the switch. This CLI is not available on the DPU`**</font>
+```
````
'Not available' means this CLI will be removed from DPU SONiC? for the above "show platform temperature" command, it said that there is also no thermal info on the DPU, shall we also disable it?

rameshraghupathy replied:
This is a sample use case.  Depending on the implementation vendor can decide to have this CLI or not. If there are no thermal sensors on the DPU  "show platform temperature" can also be disabled.

keboliu commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,947 @@
+DPU3            2023_10_02_17_23_46     Host Reset DPU              Sun 02 Oct 2023 05:23:46 PM UTC     N/A     N/A
+DPU2            2023_10_02_17_20_46     reboot                      Sun 02 Oct 2023 05:20:46 PM UTC     admin   User issued 'reboot'
+
+"show reboot-cause history <module-name>"
```
Is <module-name> option only for DPU modules? Is it possible to specify 'switch' as the module-name?

keboliu commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,947 @@
+* The "show chassis modules status" is an existing CLI but extended to include the status of all DPUs and switch. <font>**`Executed on the switch. This CLI is not available on the DPU.`**</font>
+```
+root@sonic:~#show chassis modules status                                                                                      
+Name        Description         ID         Oper-Status     Admin-Status     Serial
````
Why must the 'Serial' column be displayed in this command?

rameshraghupathy replied:
This is an existing common CLI and the serial is there.  So don't want to make any changes to it.

keboliu commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,947 @@
+Ethernet240                                  4,5,6,7     100G   9100    N/A   etp30a  routed    down       up     N/A         N/A
+Ethernet244                                  0,1,2,3     100G   9100    N/A   etp30b  routed    down       up     N/A         N/A
+Ethernet248                                8,9,10,11     100G   9100    N/A   etp31a  routed    down       up     N/A         N/A
+Ethernet252                              12,13,14,15     100G   9100    N/A   etp31b  routed    down       up     N/A         N/A
```
will the internal DPU ports name be aligned with the show command output which executed from DPU? I am asking because I see the name is different with the below command executed from DPU.

rameshraghupathy replied:
Fixed it.  Thanks for mentioning this!

keboliu commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,947 @@
+```
+#### Chassis/DPU power on sequence
+* The chassis is powered up and the host is booting up.
+* PMON gets DB notification and invokes ModuleClass API set_admin_state(self, up)
````
Which daemon will listen to the DB notification and manage the DPU?

oleksandrivantsiv commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,947 @@
+    | REBOOT_CAUSE_HOST_RESET_DPU | SW | Host lost DPU - Try resetting DPU |
+    | REBOOT_CAUSE_HOST_POWERCYCLE_DPU | SW | Host lost DPU - Power cycled DPU |
+    | REBOOT_CAUSE_SW_THERMAL |	SW | Switch software Powered Down DPU due to DPU temperature failure |
+    | REBOOT_CAUSE_DPU_SELF_REBOOT | SW | DPU Software reboots the DPU |
```
SONiC already has a list of predefined reboot causes: https://github.com/sonic-net/sonic-platform-common/blob/56921d80e42971f2579115f26c5194b9e0ecbc05/sonic_platform_base/chassis_base.py#L20
We need to try to reuse this list for the DPU as well. If from your point of view, it misses causes let's extend the list with the additional causes, rather than defining a completely new list just for DPU.

rameshraghupathy replied:
You are right that's the idea. I explicitly called out that these will be added to existing list of "Possible reboot causes" under ChassisBase Class.

manapalai reviewed (commented):
can we add show counter/stats command for DPU ports, will there be any change in output?

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,806 @@
+    * Potential consumer: Switch CLIs, Utils (install/repair images), HA, LB, Life Cycle Manager 
+    * Use cases: Debuggability, error recovery (reset, power cycle) and fault management, consolidated view of Switch and DPU state
+
+* ChassisStateDB Schema for DPU_STATE
```
We have two levels for health data.  This is the minimal viable data set to represent the DPU_STATE and the consumers for this could be HA, LB.  We also support the high volume not so critical health data using the same existing framework as you mentioned.

oleksandrivantsiv replied:
For HA and other features, it doesn't matter the exact reason why the DPU is down. It is enough to know if it is up or down. All the rest is extra information that is useful only for the operator to debug the reason why the DPU is down.

I suggest keeping in the DPU state table only the oper state. And put all other information into the system health as we agreed.

rameshraghupathy commented:
> I have several general comments on the design:
> 
> 1. There is no definition of all CLI commands that should be used to control the DPUs.
> 2. There are no flow diagrams that describe relations between the CLI commands, databases, and the platform API. It is impossible to understand how PMON will work based on the available information in this document.
> 3. There is no sonic-mgmt test plan for the proposed new functionality. The test plan is required to understand the scope of the testing and coverage of the new functionality.

Provided the full list of new and modified APIs
Added flow diagrams for power sequencing.  The CLI - DB sequence remains the same 
The sonic-management test plan will be  separate document

oleksandrivantsiv commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,941 @@
+    Key: "CHASSIS_MODULE_TABLE|DPU0"
+    "CHASSIS_MODULE_TABLE|DPU0":{
+        "value": {
+            "admin_status":"up",
```
Why do we need an admin state in State DB? I don't see it used in any of the flows.

rameshraghupathy replied:
That's a typo.  Good catch!  It should be ConfigDB and the table will be as shown below
  "CHASSIS_MODULE|DPU4": {
    "expireat": 1693437132.8384774,
    "ttl": -0.001,
    "type": "hash",
    "value": {
      "admin_status": "down"
    }
  },

oleksandrivantsiv commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,941 @@
+* The host upon ack or timeout removes the DPU port from the midplane bridge.
+* When the smartswitch is completely shutdown, it will shutdown all the DPUs first during the pre-shutdown phase of the switch. Then shutdown the switch
+*  DPU power off sequence diagram
+<p align="center"><img src="./images/dpu-pwr-off-seq.svg"></p>
```
This is my understanding of the flow for the individual DPU shutdown:
![smart-switch-pmon-power-off drawio](https://github.com/sonic-net/SONiC/assets/17766225/8532ec24-0216-46a5-b71d-e94986039b14)

rameshraghupathy replied:
Yes, however the role of GNOI needs to be confirmed yet.

rameshraghupathy replied:
Yes, your understanding is correct

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,941 @@
+* PMON gets DB notification and invokes ModuleClass API set_admin_state(self, up)
+* The host powers up the DPU (TBD: GNOI). NPU-DPU (GNOI) reboot workflow will be captured in another document
+* DPU boots up and attaches itself to the midplane.
+* Once SONiC is up the state progression is updated for every state transition on the DPU_STATE table in the chassisStateDB
```
These are all fine, but like to iterate that the minimal set of DPU_STATE will be updated into the ChassisStateDB by the DPUs as state changes.  This was requested primarily by the HA team as well.  Are you saying that this data should be populated by the DPUs into the StateDB instead of the ChassisStateDB?  If this is what you are requesting please let us knwo.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,941 @@
+<p align="center"><img src="./images/dpu-pwr-off-seq.svg"></p>
+
+#### DPU reboot sequence - reboot command issued on the DPU console
+*  The DPU reboot sequence is similar to the above and the the sequence diagram is self explanatory.
```
Agree and this is how it should work. However, it was not confirmed that the GNOI will only be used for pre-shutdown.

rameshraghupathy replied:
Yes, this is the sequence.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,947 @@
+ ...
+ ...
+
+### These are internal DPU ports ###
```
It is the same as other ports

manapalai replied:
However the 'type' field should show something like 'DPU Internal', check with @prgeor

rameshraghupathy replied:
Agree. Done

rameshraghupathy commented:
Hi Mana, Alarm and Syslog remain the same and there won't be new functions.

vvolam commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,924 @@
+
+    Returns:
+        A bool value, should return True if module is reachable via midplane
+```
````
As per our meeting discussion, are we considering implementing an API to retrieve the timeout value configured for errors that occur during the shutdown or boot-up sequence?

rameshraghupathy replied:
Not sure if I understood the question right. "is_midplane_reachable" should always return a bool.  For example the ping timeout is a well defined timeout and when that happens the platform will return  False and when it succeeds the platform will return True.

vvolam commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+* DPU: Maintenance, Critical alarm, Service migration, RMA
+#### DPU shutdown sequence
+* There could be two possible sources for DPU shutdown. 1. A configuration change to DPU "admin_status: down" 2. The GNOI logic can trigger it when it fails to communicate with the DPU.
+* The host sends a GNOI signal to gracefully shutdown the DPU.  Upon timeout the host may trigger the switch PMON to shutdown the DPU. NPU-DPU (GNOI) soft reboot workflow will be captured in another document.
```
Should we include a segment addressing ungraceful shutdown or boot-up sequences to ensure comprehensive coverage?

rameshraghupathy replied:
This is vendor specific.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+The specialized DPUs when built into a regular switch, can provide such a capability, which is being referred as SmartSwitch.
+Platform monitor PMON in SONiC is a container responsible for chassis management functions to ensure proper operation of the devices and peripherals in the chassis there by ensuring the proper operation of the product. 
+
+The typical lifecycle of a product involves the following stages.
```
In fact it refers to any network switch product. Updated.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+The following sub-tasks are performed under each stage
+* Onboarding
+    * Boot, Shutdown, Power Cycle
+    * Rest, PCIe-Reset
```
Thanks, Fixed it.

rameshraghupathy commented on doc/smart-switch/pmon/images/dpu-shutdown-seq.svg:
yes

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+#### DPU cold startup Sequence
+* The chassis is powered up and the host is booting up.
+* The switch PMON is registered with the configDB state change handler.
+* DPU "admin_status: up" is set in the configDB.
```
Thanks! When not explicitly configured the default is "up".  Updated.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+* There could be two possible sources for DPU shutdown. 1. A configuration change to DPU "admin_status: down" 2. The GNOI logic can trigger it when it fails to communicate with the DPU.
+* The host sends a GNOI signal to gracefully shutdown the DPU.  Upon timeout the host may trigger the switch PMON to shutdown the DPU. NPU-DPU (GNOI) soft reboot workflow will be captured in another document.
+* In the first option the "admin_status: down" configDB status change event will send a message to the switch PMON.
+* The switch PMON will invoke the platform API to shutdown the DPU
```
The switch PMON will invoke the module class API "set_admin_state(self, up):" with the state being "down" and the platform in turn will call its API to gracefully shutdown the DPU.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+* In the first option the "admin_status: down" configDB status change event will send a message to the switch PMON.
+* The switch PMON will invoke the platform API to shutdown the DPU
+* The DPU upon receiving the shutdown message will do a graceful shutdown and send an ack back.
+* The switch upon receiving the ack or timeout will remove the DPU from the bridge and PCIe tree.
```
The DPU upon receiving the shutdown message will do a graceful shutdown and send an ack back. The DPU graceful shutdown is vendor specific. The DPU power will be turned off after the graceful shutdown. In case of timeout the platform will force power down.
* The switch upon receiving the ack or on a timeout will remove the DPU from the bridge and PCIe tree.

rameshraghupathy replied:
The shutdown sequence will be the same as in reboot until all the services have been gracefully shutdown. At the end of graceful shutdown the GNOI logic will send a notification to invoke the platform API to power down the device.
When the GNOI will distinguish between the reboot and shutdown when the request is made.
If required will revisit this when the GNOI is completed.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+
+    | DPU Reboot Cause | HW/SW | End_User_Message_in_DPU_STATE |
+    | --- | --- | --- |
+    | REBOOT_CAUSE_POWER_LOSS | HW | Power failure |
```
Done

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+    | REBOOT_CAUSE_HOST_RESET_DPU | SW | Host lost DPU - Try resetting DPU |
+    | REBOOT_CAUSE_HOST_POWERCYCLE_DPU | SW | Host lost DPU - Power cycled DPU |
+    | REBOOT_CAUSE_SW_THERMAL |	SW | Switch software Powered Down DPU due to DPU temperature failure |
+    | REBOOT_CAUSE_DPU_SELF_REBOOT | SW | DPU Software reboots the DPU |
```
This is neither user initiated nor switch triggered. An example could be a reboot of the DPU by itself for not any known cause.

rameshraghupathy replied:
will remove this cause

rameshraghupathy replied:
will use only the reboot causes reported by the DPU and will not additional ones.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+    | REBOOT_CAUSE_SW_THERMAL |	SW | Switch software Powered Down DPU due to DPU temperature failure |
+    | REBOOT_CAUSE_DPU_SELF_REBOOT | SW | DPU Software reboots the DPU |
+
+Every time a DPU is powered up, platform can fetch this information from the DPU during chassisd module init via gRPC and update the Switch (NPU) side STATE_DB. The schema for it is an extension of the existing schema as shown.  The CLIs will use the "device" filed to identify the device.
```
During the early conversation on PMON-HLD a request was made to make the DPU reboot-cause to be available on the Switch ChassisStateDB.  Subsequently it was decided that the reboot-cause history CLI on the Switch would fetch the data from the DPUs and provide a consolidated view of the entire system.  This is about "item 3. NPU displaying reboot-cause for each DPU".  The DPUs would update this on the Switch ChassisStateDB, as per the schema mentioned.  If it is not urgent we can implement this in phase two.

rameshraghupathy replied:
1. Each DPU will update its reboot cause history in the Switch ChasissStateDB upon boot up.  The recent reboot-cause can be derived from that list of reboot-causes.

2. Sample DB schema:
  "REBOOT_CAUSE|2024_01_26_21_49_02": {
    "expireat": 1695096265.7131581,
    "ttl": -0.001,
    "type": "hash",
    "value": {
      "cause": "Unknown",
      "comment": "N/A",
      "device": "DPU5",
      "time": "2024_01_26_21_49_02",
      "user": "N/A"
    }

3. The get_reboot_cause will return the current reboot-cause of the module.
4. For storing the history will follow the existing mechanism.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+### These are internal DPU ports ###
+...
+...
+Ethernet-BP0                                  4,5,6,7     200G   9100    N/A   dpu-0  routed    down       up     Internal-DPU  N/A
```
We will use the role type as well.  Thanks.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,924 @@
+* If the vendor wants to add additional information specific to the platform it can be done using the user-defined checkers:
+* The DPU specific interrupt events can be part of this.
+```
+show system-health monitor-list <module-name>
````
Done

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+Offline: dpu_midplane_link_state or dpu_booted_state is down
+Partial Online: dpu_midplane_link_state is up and dpu_booted_state is up and dpu_control_plane_state or dpu_data_plane_state is down
+
+root@sonic:~#show system-health DPU all  
```
There are two parts to the state detail. 1. The midplane state 2. the dpu states (booted, control plane state, data plane state).  The midplane state has to be updated by the switch side pcied. The dpu states will be updated by the DPU (redis client update) on the switch ChassisStateDB.  The get_state_info() API in the moduleBase class will fetch the contents from the DB.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,924 @@
+DPU3        2023_10_02_17_23_46     Host Reset DPU                  Sun 02 Oct 2023 05:23:46 PM UTC     N/A     N/A
+```
+#### 3.4.4 Chassis Module Status
+* The "show chassis modules status" is an existing CLI but extended to include the status of all DPUs and switch. <font>**`Executed on the switch. This CLI is not available on the DPU.`**</font>
````
It exists in 3.4.5 
Online : All states are up
Offline: dpu_midplane_link_state or dpu_booted_state is down
Partial Online: dpu_midplane_link_state is up and dpu_booted_state is up and dpu_control_plane_state or dpu_data_plane_state is down

oleksandrivantsiv commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,951 @@
+#### 3.1.3 NPU to DPU data port mapping
+platform.json of NPU/switch will show the NPU to DPU data port mapping. This will be used by services early in the system boot. 
+```
+{
````
We agreed to have the following schema for this:
```
"DPUs" : [
    {
      "dpu0": {
                "interface": {"Ethernet224": "Ethernet0"}
       }
    },
    {
       "dpu1": {
                "interface": {"Ethernet232": "Ethernet0"}
        },
    },
    .
    .
    {
       "dpuX": {
                "interface": {"EthernetX": "EthernetY"}
        }
    }
]
```
Having the mapping under the `interface` key will simplify the implementation as there will be multiple attributes for each DPU.
![image](https://github.com/sonic-net/SONiC/assets/17766225/a32f57d2-2429-487c-ad7c-7abb978a60b9)

rameshraghupathy replied:
Fixed it is as shown here .

oleksandrivantsiv reviewed (changes requested):
as commented (minor change requrest)

prgeor reviewed (commented):
@rameshraghupathy Please add a section for
1. Console access of DPUs
2. Firmware upgrade of different DPU components.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,924 @@
+
+#### 3.4.1 Reboot Cause
+* There are two CLIs "show reboot-cause" and "show reboot-cause history" which are applicable to both DPUs and the Switch. However, when executed on the Switch the CLIs provide a consolidated view of reboot cause as shown below.
+* The switch will fetch the reboot-cause history from each of the DPUs as needed when the CLI is issued on the switch.
```
Done and updated the doc.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,951 @@
+#### DPU shutdown sequence
+* There could be two possible sources for DPU shutdown. 1. A configuration change to DPU "admin_status: down" 2. The GNOI logic can trigger it.
+* The GNOI server runs on the DPU even after the DPU is shutdown.
+* The host sends a GNOI signal to shutdown the DPU. The DPU does a pre-shutdown and sends an ack back to the host. The pre-shutdown is vendor specific.
```
Meant to say the actual power down sequence (turning off the power) is vendor specific.  Anyhow removed that statement.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,951 @@
+
+### 3.3   Midplane Interface
+A typical modular chassis includes a midplane-interface to interconnect the Supervisor & line-cards. When DPU card or the Switch boots and as part of its initialization, midplane interface gets initialized.
+* By default smartswitch midplane IP address assignment will be done using internal DHCP.
```
It is a couple of lines below this. Now moved it up.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,951 @@
+* Platform collects fan related data such as presence, failure and then applies fan algorithm to set the new fan speed
+* The north bound CLI/Utils/App use DB data to ”show environment”, ”show platform temp” show platform fan”
+
+Thermal management sequence diagram
```
Actually how the DPU sensor data gets to the Switch StateDB is vendor implementation specific.  The sequence diagram is little off. Updated it.  The existing "TEMPERATURE_INFO" schema is the one used for this purpose. Added that also to the docuemnt for clarity.

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,951 @@
+#### DPU_STATE Use Case
+* High Availability, Load Balancing, Debug, error recovery (reset, power cycle) and fault management logics will use this API
+
+get_state_info(self):
```
The understanding is that the HA logic may use this.  However, how and when will be defined in the HA document.  If needed at that time will update this section.  Also, the there was minor issue that mentioned that this is used in the DPU instead of SWITCH, which is fixed now.  The purpose of this is to keep this as a fall back option or even as the main option in case of any BFD implementation issues.

rameshraghupathy commented:
The following 4 PRs have been raised for PMON Platform Independent changes:
https://github.com/sonic-net/sonic-buildimage/pull/18676
https://github.com/sonic-net/sonic-platform-daemons/pull/467
https://github.com/sonic-net/sonic-platform-common/pull/454
https://github.com/sonic-net/sonic-utilities/pull/3271

oleksandrivantsiv commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,1027 @@
+* The default baud rate should be 9600
+
+### 3.6 Firmware Upgrade
+* The existing fwutil will be extended to the DPUs
```
What is new for the SmartSwitch here? The utility should remain unchanged, but the vendor plugin should be extended to support the DPU FW upgrade.

rameshraghupathy replied:
Agree, that's the plan

oleksandrivantsiv replied:
Can you please add a note about this to the document?

oleksandrivantsiv commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -946,6 +982,55 @@ root@sonic:/home/admin# show interfaces status
+* Provide a toggle option to access the BMC console in order to recover the CPU in case of a failure
+* Once inside the switch, the DPUs should be accesssible from it with an utility. Platforms can use this utility to warp the underlying generic utilitites such as picocom or rconsole.
+    * Ex: xconsole -b <baud-rate> -m <module-name>
+* The default baud rate should be 9600
```
As this is a serial console from NTP to the DPU there should not be restrictions for the baud rate. The baud rate can remain up to the vendor.

rameshraghupathy replied:
The baud rate is programmable, but Microsoft has requested the default baud rate to be set to 9600.  So I called out the default baud rate.

rameshraghupathy replied:
If Price is ok removing the default baud rate I can do it.

oleksandrivantsiv replied:
@prgeor do we need to limit the baud rate? The serial connection between NPU and DPU is internal to the system, so I assume any baud rate is good while both sides are aligned.

rameshraghupathy replied:
We don't have to. My understanding is that the existing data centers have all their equipments connecting at 9600.  So, setting the default of the smartswitch baud rate to 9600 will not require any additional script change at the datacenter end. I tried my best to clarify your question.  However, I removed that line as well :)

rameshraghupathy commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,1027 @@
+### 3.5 Console Management
+The console access to smartswitch needs to support DPU access in addition to the switch.
+* By default the management port should be connected to the switch CPU console
+* Provide a toggle option to access the BMC console in order to recover the CPU in case of a failure
```
Done

r12f commented on doc/smart-switch/pmon/smartswitch-pmon.md:
````diff
@@ -0,0 +1,1025 @@
+```
+
+3. Each DPU has to store the health data in its local DB and should provide it to the switch.
+* When the "show system-health ..." CLI is executed on the switch, the "UserDefinedChecker" class will collect the data and feed it to the CLI. It is up to the platform on how this is done.  However, for faster access store it in the switch ChassisStateDB.
````
as chatted offline, let's add a notification channel between some component in PMON to hamgrd for notifying the state being changed, as these are critical health signal HA is relying on.

prgeor replied:
@r12f hamgrd can subscribe to the redis db table events to get these notifications

rameshraghupathy replied:
Sure, I'll add another document to detail out the interactions between the hamgrd and PMON dpu_state change handler.

prgeor commented on doc/smart-switch/pmon/smartswitch-pmon.md:
```diff
@@ -0,0 +1,1025 @@
+        ”dpu_control_plane_reason": ”containers restarting",
+        ”dpu_data_plane_state": ”down",
+        ”dpu_data_plane_time": ”timestamp",
+        ”dpu_data_plane_reason": ”Pipeline failure",
```
@r12f Redis DB provide subscription to table change events. hmgrd can subscribe to these tables

rameshraghupathy replied:
Agree, Also I'll add another document to cover the hamgr interactions and event handler notifications
